### PR TITLE
Breakpoints redesign stage 2

### PR DESCRIPTION
**Pull request by kof**
## Description

- Canvas behavior should be exactly as in Webflow now - responsive canvas.
- No constraints for resizing, you can switch to different breakpoints when you change the canvas width just like in preview

## Not done in this PR

- [ ] breakpoints dropdown menu redesign
- [ ] on-canvas resizable handles
- [ ] flash-animate breakpoints list when breakpoint is switched through resize
- [ ] flash-animate left and right borders when breakpoint is switched
- [ ] gravity of resize handle dragging when reaching next breakpoint
- [ ] vertical labels with @media example that matches current breakpoint
- [ ] migrate min-width 0 breakpoints to base?
- [ ] hide outline while changing size
- [ ] ensure there is never both min and max in one media

## Todo

- [x] draggable positioning calc
- [x] shortcuts for zoom
- [x] sync width and selected breakpoint
- [x] remove scale
- [x] fix mobile-first breakpoints
- [x] tabbable toolbar items
- [x] icon for base
- [x] fix flicker
- [x] After breakpoint delete/add CascadeIndicator doesn't refresh
- [x] fix when there is no base, min-width 0 should have still a full width

## Steps for reproduction

1. click button
3. expect xyz

## Code Review

- [ ] hi @TrySound, @istarkov  , I need you to do
  - conceptual review (architecture, feature-correctness)
  - detailed review (read every line)
  - test it on preview

## Before requesting a review

- [ ] made a self-review
- [ ] added inline comments where things may be not obvious (the "why", not "what")

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio-builder/blob/main/apps/builder/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env.example` and the `builder/env-check.js` if mandatory

### CODE REVIEW

File: packages/css-engine/src/core/find-matching-media.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import { compareMedia } from "./compare-media";
+import { matchMedia } from "./match-media";
+import type { MediaRuleOptions } from "./rules";
+
+// Find media rule that matches the given width when rendered.
+export const findMatchingMedia = <Media extends MediaRuleOptions>(
+  media: Array<Media>,
+  width: number
```
Comment (TrySound) [inline packages/css-engine/src/core/find-matching-media.ts:8]: Isn't it the same as match media?

**Comment by taylornowotny** [comment]
When opening any breakpoint the canvas goes to max size and there's a message saying:

 "Please increase the canvas width.
"Mobile portrait" breakpoint maximum width is 479px."

I can resize with the menu but it's jumpy and flashes at max size at each interval.

The intended behavior from the docs is:
> "Switching breakpoints with the toggles will change the canvas width to the smallest available size for that breakpoint
> - if bp has a min-width, the min-width is the smallest size
> - if bp has a max-width, then the smallest size for that bp is the max-width of the next smallest bp +1 pixel
> - This behavior is same as Webflow so you can use them as a reference
> - Reasoning: A best practice for the user is to apply styles to the smallest width of a given bp, bc if content is designed for the smaller end then it will display properly on the larger end, but not necessarily the other way around."

**Comment by taylornowotny** [comment]
Screenshot 2023-04-07 at 9 29 48 AM

+ can we give this button a min-width of 121px so it doesn't resize when the canvas width value goes from 4 digits to 3 digits & back? The movement is a bit distracting.

**Comment by kof** [comment]
> can we give this button a min-width of 121px so it doesn't resize when the canvas width value goes from 4 digits to 3 digits & back? The movement is a bit distracting.

I was also thinking to hide the scale value when its 100%, since most of the time its not useful, webflow hides it too, but then I saw the breakpoints selector toolbar jumping around and removed this logic, with min-width we could potentially work this around  and make it stay

**Comment by taylornowotny** [comment]
canvas setting button looks great now when resizing!

**Comment by istarkov** [comment]
By default project opens with 0 width canvas
https://webstudio-builder-59fpf7vb1-webstudio-is.vercel.app/builder/71f88d3e-9231-489e-8ba0-8adc10215335
So nothing seen at all

**Comment by istarkov** [comment]
If I set width for example 354px
image
Then switch on 768 clicking here
image
Then click 0
Canvas width is 0 again

**Comment by istarkov** [comment]
If I set canvas width 
image
Then resize window it resets to defaults, 
image

IMO precents should be changed but not pixel size, if user (like me) has changed size explicitly

File: apps/builder/app/builder/features/breakpoints/cascade-indicator.tsx [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  return (
+    <>
+      <div
+        className={cascadeIndicatorStyle()}
+        data-direction="left"
+        style={indicatorStyle.left}
+      />
+      <div
+        className={cascadeIndicatorStyle()}
+        data-direction="right"
```
Comment (istarkov) [inline apps/builder/app/builder/features/breakpoints/cascade-indicator.tsx:121]: In UI I see single indicator, what the reason to have left and right?
Any comments why we need both would be good.

**Comment by istarkov** [comment]
After breakpoint delete/add CascadeIndicator doesn't refresh
image

File: apps/builder/app/builder/features/breakpoints/breakpoints-selector.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+              <ToolbarToggleItem
+                variant="subtle"
+                ref={breakpoint.id === selectedBreakpoint.id ? ref : undefined}
+                value={breakpoint.id}
+                key={breakpoint.id}
+              >
+                {breakpoint.minWidth ??
+                  breakpoint.maxWidth ??
+                  (breakpoint.id === selectedBreakpoint.id ? (
+                    <BpStarOnIcon size={22} />
```
Comment (istarkov) [inline apps/builder/app/builder/features/breakpoints/breakpoints-selector.tsx:44]: Probably I'm doing something wrong but how to see Star icon on the indicator? What I need to do?

File: apps/builder/app/builder/features/breakpoints/group-breakpoints.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { compareMedia } from "@webstudio-is/css-engine";
+import { isBaseBreakpoint } from "~/builder/shared/breakpoints";
+//import type { Breakpoint } from "@webstudio-is/project-build";
```
Comment (istarkov) [inline apps/builder/app/builder/features/breakpoints/group-breakpoints.ts:3]: remove

File: apps/builder/app/builder/features/breakpoints/group-breakpoints.test.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+describe("Group breakpoints", () => {
+  test("for ui", () => {
+    const initial = [
+      { minWidth: 1920 },
+      { minWidth: 1440 },
+      { minWidth: 1280 },
+      {},
+      { maxWidth: 991 },
+      { maxWidth: 767 },
+      { maxWidth: 479 },
```
Comment (istarkov) [inline apps/builder/app/builder/features/breakpoints/group-breakpoints.test.ts:13]: Are we going to mix breakpoints?

File: apps/builder/app/builder/features/breakpoints/group-breakpoints.test.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    const grouped = [
+      { minWidth: 1920 },
+      { minWidth: 1440 },
+      { minWidth: 1280 },
+      {},
+      { maxWidth: 991 },
+      { maxWidth: 767 },
+      { maxWidth: 479 },
+    ];
+    expect(groupBreakpoints(initial)).toStrictEqual(grouped);
```
Comment (istarkov) [inline apps/builder/app/builder/features/breakpoints/group-breakpoints.test.ts:24]: I don't get this test at all, as it looks like you test that initial array is equal to grouped. Having that initial === grouped it makes zero sense.

File: apps/builder/app/builder/features/breakpoints/group-breakpoints.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+export const groupBreakpoints = <
```
Comment (istarkov) [inline apps/builder/app/builder/features/breakpoints/group-breakpoints.ts:5]: Doesn't look like a group, but reorder/order as you don't have any groups here.

File: apps/builder/app/builder/features/workspace/workspace.tsx [hunk 8]
```diff
@@ -45,6 +39,37 @@ const useSetWorkspaceRect = () => {
...
+) => {
+  let canvasHeight;
+  let canvasLeft;
+
+  if (workspaceRect?.height) {
+    canvasHeight = workspaceRect.height / (scale / 100);
+  }
+
+  if (workspaceRect?.width && canvasWidth) {
+    canvasLeft = Math.max((workspaceRect.width - canvasWidth) / 2, 0);
```
Comment (istarkov) [inline apps/builder/app/builder/features/workspace/workspace.tsx:55]: I have a feeling that canvasLeft>0 only if transform scale === 1, 
that allows to not calculate left position but use for canvas positioning ordinary Flex box with justify-content: center. 
Less calculus is always better

File: packages/css-engine/src/core/find-applicable-media.test.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+const media = [
+  {},
+  { maxWidth: 991 },
+  { maxWidth: 767 },
+  { maxWidth: 479 },
+  { minWidth: 1280 },
+  { minWidth: 1440 },
+  { minWidth: 1920 },
+];
```
Comment (istarkov) [inline packages/css-engine/src/core/find-applicable-media.test.ts:12]: What if maxWidth and minWidth would have intersections? How grouping etc would work too

File: packages/css-engine/src/core/find-applicable-media.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
+import { compareMedia } from "./compare-media";
+import { matchMedia } from "./match-media";
+import type { MediaRuleOptions } from "./rules";
+
+// Find media rule that matches the given width when rendered.
+export const findApplicableMedia = <Media extends MediaRuleOptions>(
+  media: Array<Media>,
+  width: number
+) => {
+  const sortedMedia = media
```
Comment (istarkov) [inline packages/css-engine/src/core/find-applicable-media.ts:10]: Looks like you need `groupBreakpoints` here too?

**Comment by kof** [comment]
@istarkov 

> By default project opens with 0 width canvas
> https://webstudio-builder-59fpf7vb1-webstudio-is.vercel.app/builder/71f88d3e-9231-489e-8ba0-8adc10215335
> So nothing seen at all

I am getting unexpected server error on this one, how do I see it?

**Comment by kof** [comment]
> I am getting unexpected server error on this one, how do I see it?

Found an example of 0 canvas width, fixing

**Review by istarkov — APPROVED** [review]

File: apps/builder/docs/test-cases.md [hunk 11]
```diff
@@ -65,7 +65,7 @@
 
    - Clicking on different breakpoints resizes the canvas
    - Canvas width slider lets you resize canvas only within its bounds
-   - Zoom slider works
+   - Scal slider works
```
Comment (TrySound) [inline apps/builder/docs/test-cases.md:68]: scal

File: apps/builder/app/shared/shortcuts/shortcuts.ts [hunk 12]
```diff
@@ -93,4 +54,6 @@ export const useSharedShortcuts = () => {
     { enableOnFormTags: true, enableOnContentEditable: true },
     []
   );
+
+  useHotkeys("esc", escapeSelection, { enableOnContentEditable: true }, []);
```
Comment (TrySound) [inline apps/builder/app/shared/shortcuts/shortcuts.ts:58]: Ivan removed it from here. Is it rebase?

File: apps/builder/app/builder/shared/nano-states/index.ts [hunk 13]
```diff
@@ -24,3 +24,19 @@ export const projectContainer = atom<Project | undefined>();
...
+  [canvasWidthContainer, workspaceRectStore],
+  (canvasWidth, workspaceRect) => {
+    if (
+      canvasWidth === undefined ||
+      workspaceRect === undefined ||
+      canvasWidth <= workspaceRect.width
+    ) {
+      return 100;
+    }
+    return parseFloat(((workspaceRect.width / canvasWidth) * 100).toFixed(2));
```
Comment (TrySound) [inline apps/builder/app/builder/shared/nano-states/index.ts:40]: Number.parseFloat

File: apps/builder/app/builder/shared/breakpoints/is-base-breakpoint.ts [hunk 14]
```diff
@@ -0,0 +1,4 @@
+export const isBaseBreakpoint = (breakpoint: {
```
Comment (TrySound) [inline apps/builder/app/builder/shared/breakpoints/is-base-breakpoint.ts:1]: shared/breakpoints.ts?

**Review by TrySound — APPROVED** [review]

File: apps/builder/app/builder/features/style-panel/style-panel.tsx [hunk 15]
```diff
@@ -7,7 +7,7 @@ import {
 } from "@webstudio-is/design-system";
 import type { Instance } from "@webstudio-is/project-build";
 import type { Publish } from "~/shared/pubsub";
-import { selectedBreakpointStore } from "~/shared/nano-states/breakpoints";
+import { selectedBreakpointStore } from "~/builder/shared/breakpoints/stores";
```
Comment (TrySound) [inline apps/builder/app/builder/features/style-panel/style-panel.tsx:10]: Why reexport from index if import from stores?

**Pull request merged**